Porting action workflow to Python 3.9.10 (ScriptComponent in V8)

9and3 opened the pull request:
Hello!


This is the first working draft to port the compas action componentizer workflow to Python 3.x.  in Grasshopper v8 (see issue #12). I tested it in CI and on other local projects and it works well, but I can't promise it to be fully bug free :) Looking forward to recive feedback to integrate and thanks again for this great workflow! 🙏


![image](https://github.com/compas-dev/compas-actions.ghpython_components/assets/50238678/d5c4f6cd-4540-46bc-9855-bc7e9a89fcf9)

🐍🐍🐍


## changelog
* porting action and componentizer from ipy to cpy (pythonnet mainly)
* some guids changed and added new guids for typehints (extrusion and pointcloud) +modified the old ones for str, float, ghdoc, and none.
* implemented the new `Script` logic in `componentizer.py`
* the `metadata.json` is cleaned off the ironpython outdated entries in the serialization and updated with the new ones


## discussions
- [ ] This version is serializing all components into CPython components, it does not distinguish python2.x from 3.x. Hence legacy ironpython code will fail in some cases with this action. *I wonder if both need to be recognized and serialized differently?*
- [ ] not important: I the old `out` parameter is not shown by default. Especially now with the new ScriptComponent it seems that the `problem` tab is taking up its role. As a python coder in Rhino I am kinda affectionate to it, so maybe it could be set as an option in `metadata.json` to wether or not showing it in the component (?).
- [ ] haven't tested modules import (yet)

9and3 committed:
WIP: setting testing env for ci

9and3 committed:
WIP: testing on cpy_dev ci branch

9and3 committed:
WIP: added also main for ci testing

9and3 committed:
WIP-CAP: ironpy to cpy compiling olc cpy

9and3 committed:
WIP: notes on xml, next it's testing deserialization on ghuser cpy

9and3 committed:
WIP: working on deseriaization

9and3 committed:
WIP: test commit

9and3 committed:
FIX: got rid of test in readme

9and3 committed:
WIP: compilationg works but component is C#, why?

9and3 committed:
WIP-CAP: changed all output guid to no_hint (out parameter is not defualt)

9and3 committed:
FIX: modified and verified guid typeHint compatibility

9and3 committed:
CAP: working marshalling python code

9and3 committed:
MILESTONE: working, rearraning structure for PR

9and3 committed:
WIP: preparing restructure for pr - test commit

9and3 committed:
WIP: testin on local ci

9and3 committed:
FIX: wrong action path from build solved

9and3 committed:
Merge branch 'restructure_cpy'

gonzalocasas commented on action.yml:
```diff
@@ -1,3 +1,5 @@
+# This action was implemented from the [compass componentizer](https://github.com/compas-dev/compas-actions.ghpython_components/tree/main). We retain no credit for the code, and only modified it to work with our own repository.
```
Is this an internal comment? Does it still make sense?

9and3 replied:
Yep, not needed.

gonzalocasas commented on componentize.py:
```diff
@@ -238,36 +243,35 @@ def create_ghuser_component(source, target, version=None, prefix=None):
         params.SetGuid(
-            "InputId", i, System.Guid.Parse("84fa917c-1ed8-4db3-8be1-7bdc4a6495a2")
+            "InputId", i, System.Guid.Parse("08908df5-fa14-4982-9ab2-1aa0927566aa")
         )
     params.SetInt32("OutputCount", len(outputParam))
     for i, _po in enumerate(outputParam):
         params.SetGuid(
-            "OutputId", i, System.Guid.Parse("8ec86459-bf01-4409-baee-174d0d2b13d0")
+            "OutputId", i, System.Guid.Parse("08908df5-fa14-4982-9ab2-1aa0927566aa")
```
Input and Output guids are matching?

9and3 replied:
Yes. From what I understood the `"OutputId" can have two possible GUIDs:
- A) `08908df5-fa14-4982-9ab2-1aa0927566aa` is the one here. It corresponds to the new  `ScriptParam`  and it works with all output variables and GHtypes.
- B) `3ede854e-c753-40eb-84cb-b48008f14fd4` is the old ghpython component's `out` parameter that streams all the console's logs to a GHstring. If given to other types of output variables the component throws a casting error.

gonzalocasas commented on componentize.py:
```diff
@@ -1,33 +1,33 @@
-GHPYTHON_SCRIPT_GUID = System.Guid("410755b1-224a-4c1e-a407-bf32fb45ea7e")
+
+GHPYTHON_SCRIPT_GUID = System.Guid("c9b2d725-6f87-4b07-af90-bd9aefef68eb")
+CPY_VER = "3.-1"
```
What does this do? Why is it `-1`?

9and3 replied:
It specify just the major version of Python to be used in the Script component without minor and patch. I tested also with the string `3.9.10` it works both ways. Would it be better to specify the version in the format `major.minor.patch`?

gonzalocasas commented on componentize.py:
```diff
@@ -1,33 +1,33 @@
 
-GHPYTHON_SCRIPT_GUID = System.Guid("410755b1-224a-4c1e-a407-bf32fb45ea7e")
+
+GHPYTHON_SCRIPT_GUID = System.Guid("c9b2d725-6f87-4b07-af90-bd9aefef68eb")
```
Is this GUID the new Script component? If so, the constant should be renamed from `GHPYTHON_SCRIPT_GUID` to `SCRIPT_COMPONENT_GUID` or similar

9and3 replied:
Modified.

gonzalocasas commented on componentize.py:
```diff
@@ -378,19 +389,20 @@ def create_ghuser_component(source, target, version=None, prefix=None):
-    print()
+
+    gh_io = os.path.abspath(gh_io)
+    gh_io = gh_io[:-4]
```
what is being chopped off of the end here?

9and3 replied:
It chops off the extension `.dll` from the path because the new `clr.AddReference()` doesn't accept it.

Maybe it is better without indexing (?):
```py
clr.AddReference(os.path.splitext(gh_io)[0])
```

gonzalocasas commented on examples/TestCpy_KitchenSink/code.py:
```diff
@@ -0,0 +1,25 @@
+    Returns:
+        result: The sum of all three values.
+"""
+from ghpythonlib.componentbase import executingcomponent as component
```
the legacy ghscript component has a VERY strange internal logic that depends on the class of the component inheriting from something called `component`. Rhino checked this using a string comparison, not type checking. That's the reason why `executingcomponent` is renamed here. Is this silly thing still true for the new script component?

9and3 replied:
Alright, thanks for the info, I was wondering at one moment why. 
Just tried and the following seems to be working without problems:
```py
from ghpythonlib.componentbase import executingcomponent
...
class MyComponent(executingcomponent):
    ....
```

9and3 replied:
mmm I see. I tried with inheritance directly from `executingcomponent` and the `ghenv.Component.Message` stops working although the ghuser component can be built and added to the canvas. If I go back to `component` everything works as expected. 

also @gonzalocasas : I do not know if it is connected but the SDK mode is not working anymore since this was a feature of the old ghpython component to activate with `IsAdvancedMode` set to True. To my limited extent of knowledge of the new ScriptComponent there is no such feature.

gonzalocasas commented on examples/TestCpy_KitchenSink/metadata.json:
```diff
@@ -0,0 +1,65 @@
+                "showTypeHints": true,
+                "scriptParamAccess": "item",
+                "wireDisplay": "faint",
+                "sourceCount": 0,
```
What is `sourceCount`?

9and3 replied:
Frankly, I don't know. It was there before and I kept it. I only know that if entryy is not serialized in the `componentizer.py`, the generated ghuser crashes once added to the canvas.

gonzalocasas reviewed (commented):
Wow! This is super cool! 
I added a bunch of comments. And there's one thing I'd like to request to change: it might be better for a smoother migration of users already using this action, that we don't remove the support for iropython, but instead add support for cpython, It could be as simple as reverting the changes on `componentize.py` and creating a `componentize-cpytyhon.py` with the new support for py3. Then, I guess the `action.yml` will need to determine which one to run based on some input.

So, by default, this action could still do ironpython, but passing a param, it could switch to build py3 components.

9and3 commented:
> Wow! This is super cool! I added a bunch of comments. 

thanks! I am going through them.

> And there's one thing I'd like to request to change: it might be better for a smoother migration of users already using this action, that we don't remove the support for iropython, but instead add support for cpython, It could be as simple as reverting the changes on `componentize.py` and creating a `componentize-cpytyhon.py` with the new support for py3. Then, I guess the `action.yml` will need to determine which one to run based on some input.
> So, by default, this action could still do ironpython, but passing a param, it could switch to build py3 components.

Got it. I'll give it a go!

9and3 committed:
WIP:integrated minor modifications from ccde review

9and3 committed:
WIP: testing reformat CI

9and3 committed:
WIP: testing reformat CI x2

9and3 committed:
WIP: testing out dual branch

9and3 committed:
WIP: testing out dual branch x2

9and3 committed:
WIP: testing out dual branch x2

9and3 committed:
WIP: testing out dual branc, two componentize

9and3 committed:
ADD: seperate examples folder

9and3 committed:
ADD: new main build

9and3 committed:
FIX: add parameter to action

9and3 committed:
FIX: specified different build out directories for artifacts

9and3 committed:
FIX-WIP: testing nuget in action

9and3 committed:
FIX: reversed to compas format for build workflow and action isntalls

9and3 committed:
WIP-CAP: final testing before refactoring readme instructions

9and3 committed:
WIP: retesting only one main workflow

9and3 committed:
CAP: final modifications and readme updated

9and3 committed:
FIX: forgot to set the example action to market ref

9and3 commented:
Hello! 

I added the dual functioning of the `action.yml` and updated both examples and readme. Technically it should have no conflict with the current version since the parameter to switch between cpython and ironpython is by default to the last one:

```yml
- uses: compas-dev/compas-actions.ghpython_components@v2
    with:
      source: components
      target: build
      interpreter: cpython  # optional, defaults to ironpython
```
I tried to update the readme and add two jobs in the `build.yml` to build both.

Let me know if there is something to modify!

gonzalocasas commented on README.md:
```diff
@@ -1,8 +1,8 @@
 # GHPython Componentizer
 
-> A github action to make Grasshopper development 164% <sup><small>[1]</small></sup> version-control friendlier and 82% more pleasant.
+> A github action to make Grasshopper development 165% <sup><small>[1]</small></sup> version-control friendlier and 83% more pleasant.
```
❤️

gonzalocasas commented on README.md:
```diff
@@ -43,28 +47,60 @@ jobs:
 Then start the script pointing it to a source and target folder, e.g.:
 
-    ipy componentize.py examples build
+    ipy componentize_ipy.py examples build
+    python componentize_cpy.py examples build
```
```suggestion
    ipy componentize_ipy.py examples/ipy build
    python componentize_cpy.py examples/cpy build
```

gonzalocasas commented on README.md:
```diff
@@ -43,28 +47,60 @@ jobs:
 Optionally, tag it with a version:
 
-    ipy componentize.py examples build --version 0.1.2
+    ipy componentize_ipy.py examples build --version 0.1.2
+    python componentize_cpy.py examples build --version 0.1.2
```
```suggestion
    ipy componentize_ipy.py examples/ipy build --version 0.1.2
    python componentize_cpy.py examples/cpy build --version 0.1.2
```

gonzalocasas commented on README.md:
```diff
@@ -43,28 +47,60 @@ jobs:
 An optional name prefix can help tell components apart from other similarly named ones:
 
-    ipy componentize.py examples build --prefix "(PACKAGE-NAME)"
+    ipy componentize_ipy.py examples build --prefix "(PACKAGE-NAME)"
+    python componentize_cpy.py examples build --prefix "(PACKAGE-NAME)"
```
```suggestion
    ipy componentize_ipy.py examples/ipy build --prefix "(PACKAGE-NAME)"
    python componentize_cpy.py examples/cpy build --prefix "(PACKAGE-NAME)"
```

gonzalocasas commented on componentize_cpy.py:
```diff
@@ -0,0 +1,405 @@
+import tempfile
+import urllib.request, urllib.parse, urllib.error
+import zipfile
+from io import StringIO
```
```suggestion
from io import BytesIO
```

gonzalocasas commented on componentize_cpy.py:
```diff
@@ -0,0 +1,405 @@
+    ghio_dll = "GH_IO.dll"
+    filename = "lib/net48/" + ghio_dll
+
+    response = urllib.urlopen("https://www.nuget.org/api/v2/package/Grasshopper/")
```
```suggestion
    response = urllib.request.urlopen("https://www.nuget.org/api/v2/package/Grasshopper/")
```

gonzalocasas commented on componentize_cpy.py:
```diff
@@ -0,0 +1,405 @@
+
+    response = urllib.urlopen("https://www.nuget.org/api/v2/package/Grasshopper/")
+    dst_file = os.path.join(target_folder, ghio_dll)
+    zip_file = zipfile.ZipFile(StringIO.StringIO(response.read()))
```
```suggestion
    zip_file = zipfile.ZipFile(BytesIO(response.read()))
```

9and3 replied:
Alright, I am leaving the `metadata.json` for the ipy examples with capital though.

gonzalocasas commented on examples/cpy/Test_KitchenSink/metadata.json:
```diff
@@ -0,0 +1,65 @@
+        "iconDisplay": 2,
+        "inputParameters": [
+            {
+                "name": "X",
```
It seems something weird happens with these names: if this is capitalized, the variable will actually be renamed:
Using capital in the name here modified the actual variable to get
```suggestion
                "name": "x",
```

gonzalocasas commented on examples/cpy/Test_KitchenSink/metadata.json:
```diff
@@ -0,0 +1,65 @@
+                "simplify": false
+            },
+            {
+                "name": "Y",
```
```suggestion
                "name": "y",
```

gonzalocasas commented on examples/cpy/Test_KitchenSink/metadata.json:
```diff
@@ -0,0 +1,65 @@
+                "simplify": true
+            },
+            {
+                "name": "Z",
```
```suggestion
                "name": "z",
```

gonzalocasas reviewed (commented):
It's looking great! I added some small feedback, after addressing those, we can merge!

9and3 committed:
REVIEW: final modifications integrated

9and3 commented:
Done! I tested it again and all is good from my side. Thanks @gonzalocasas for the feedback, this super nice workflow and for compas y'all🙏

9and3 committed:
ADD-MINOR: description is used for tooltipping the component

gonzalocasas reviewed (approved):
Excellent! I'm merging this, updating docs to reference a new ~v3~ v5 release, and releasing it! Thanks a lot!
